rholang/src/main/k/rholang.k

```diff
@@ -1,132 +1,1039 @@
 module RHO-SYNTAX
 imports SUBSTITUTION
 
-// Syntax for processes
-syntax Proc ::=
-              // Empty process
-                "Nil"
+// We need processes, including variables
+syntax Proc ::= ProcNoVars
+              | Id
+
+// We need processes, excluding variables
+syntax ProcNoVars ::=
+              // Ground terms, including the empty process and expressions
+                Ground
               // Listen
-              | "for(" Name "<-" Name "){" Proc "}" [binder]
+              | "for" "(" Names "<-" Name ")" "{" Proc "}" [binder]
               // Persistent listen
-              | "for(" Name "<=" Name "){" Proc "}" [binder]
+              | "for" "(" Names "<=" Name ")" "{" Proc "}" [binder]
               | "contract" Name "(" Name ")" "=" "{" Proc "}" [binder]
               // Send
-              | Name "!(" Proc ")"
+              | Name "!(" Procs ")"
               // Evaluate
               | "*" Name
               // New
-              | "new" Id "in" "{" Proc "}" [binder]
+              | "new" Ids "in" "{" Proc "}" [binder]
               // Parallel
               | Proc "|" Proc [left]
-              // Misc
+
+// We put "Nil" and Exp in the same syntactic category, since they are
+// structurally equivalent as ground properties
+syntax Ground ::= "Nil"
               | Exp
 
-// Syntax for Names; we allow variables to take the place of names, but note
-// that no rholang program will compile if there are free variables left in
-// the program. Thus the variable must be bound to a name of the form @Proc
-syntax Name ::= "@" Proc
+// We need names with no variables
+// (We need put the Proc through some algorithm which puts it into its normal form; will come later)
+syntax NameNoVars ::= "@" Proc
+
+// The general idea of a name can be a (bound) variable
+syntax Name ::= NameNoVars
+              | Id
+
+// Names and Processes, used specifically for sending and listening for tuples
```

Comment:
A tuple is just another process, right? Is `x!(P, Q, Z)` valid rholang? or do we send the tuple `x!((P,Q,Z))`?

Reply:
`x!(P, Q, Z)` is a process in rholang but the tuple by itself is not. The syntax is `x!(P, Q, Z)`, and not `x!((P,Q,Z))` although I don't quite know the exact philosophical interpretation of the tuple. I think of it as sending the processes grouped together and in order, able to match with a pattern of the same, but not as a process itself. Maybe @pyrocto can enlighten?

Reply:
When you read from a channel, you get a list of names, not a process:
`for (x, y, z <- chan) { ... }`
If you were only able to send a tuple, all vars would necessarily be process vars.

Reply:
Yeah, sorry, I suppose the receive would be `for( @(A, B, C) <- x)`.

If we can construct process lists which are themselves processes (and we need to for some of the compound name examples), why don't we just send those and have the matcher do the airity checking on the other side? Seems like it would simplify the syntax. Or am I not understanding something that would be lost.

Reply:
It's a compromise between always receiving names and having to put a `*` on everything when you use it as a process (e.g. `for (i <- x; j <- y; k <- z) { ret!(*i * *j * *k) }`) vs always receiving processes and having to put `@`s on everything when we use it as a name (e.g. `for (i <- @x; j <- @y; k <- @z) { @i!(1) | @j!(2) | @k!(3) }`.

Reply:
Sure, I see. It still might make sense to simplify this definition by defining a structural rule to transform the form `for ( x, y <- a )` / `a!( P, Q )` to the form `for ( @(@x, @y) <- a)` / `a!( (P, Q) )`

Reply:
`for ( @(@x, @y) <- a) { ... }` doesn't even parse: the thing in the first slot of the tuple is a process, not a name, so if you wanted a name variable, you'd have to use something like `for (@(*x, *y) <- a) { ... }` where you're trying to say that the process at each slot in the tuple is the evaluation of some name.  But the matcher looks for an actual `*` operator in the AST, so it wouldn't work because all evaluations are done before serializing.